rust/src/utils.rs

```diff
@@ -15,42 +18,100 @@ pub fn is_eligible(signature: &ParticipantTaskSignature, threshold: f64) -> bool
     } else if threshold > 1_f64 {
         true
     } else {
-        Ratio::new(
-            BigUint::from_bytes_be(&sha256::hash(signature.as_slice()).0[..])
-                .to_bigint()
-                // FIXME: can we unwrap here?
-                .unwrap(),
-            BigUint::from_bytes_be(&[255_u8; 32][..])
-                .to_bigint()
-                .unwrap(),
-        ) <= Ratio::from_float(threshold).unwrap()
+        // safe unwraps: `to_bigint` never fails for `BigUint`s
+        let numer = BigUint::from_bytes_le(sha256::hash(signature.as_slice()).as_ref())
+            .to_bigint()
+            .unwrap();
+        let denom = BigUint::from_bytes_le([255_u8; 32].as_ref())
+            .to_bigint()
+            .unwrap();
+        // safe unwrap: `threshold` is guaranteed to be finite
+        Ratio::new(numer, denom) <= Ratio::from_float(threshold).unwrap()
     }
 }
 
+/// Generate a secure pseudo-random integer. Draws from a uniform distribution over the integers
+/// between zero (included) and `max_int` (excluded).
+pub fn generate_integer(prng: &mut ChaCha20Rng, max_int: &BigUint) -> BigUint {
```

Comment:
Maybe this could go in the `crypto` module?

Reply:
i'll move it.

Reply:
i'll also move the `is_eligible()` to `crypto::sign`.